File: implementations/python/src/esque/migration/loader.py
from __future__ import annotations

import hashlib
import json
import re
import struct
from pathlib import Path
from typing import Any, cast

import yaml

from esque.migration.model import (
    MigrationFile,
    MigrationFileContents,
    MigrationFileMetadata,
    MigrationFileRequestDefinition,
)
from esque.migration.template import MigrationTemplateResolver

_FILE_NAME_PATTERN = re.compile(r"^V((\d+\.?)+)__(\w+)\.yml$")


class MigrationFileLoader:
    def __init__(self, migration_directory: str, template_resolver: MigrationTemplateResolver) -> None:
        self._migration_directory = migration_directory
        self._template_resolver = template_resolver

    def load(self) -> list[MigrationFile]:
        path = self._resolve_path(self._migration_directory)
        raw_files: list[MigrationFile] = []
        for file_path in path.iterdir():
            if not file_path.is_file():
                continue
            match = _FILE_NAME_PATTERN.match(file_path.name)
            if not match:
                continue
            raw_files.append(self._read_raw(file_path, match))

        raw_files.sort()
        self._template_resolver.validate(raw_files)
        return [self._resolve(f) for f in raw_files]

    def _resolve(self, file: MigrationFile) -> MigrationFile:
        resolved_contents = self._template_resolver.resolve_contents(file.contents)
        return MigrationFile(
            metadata=MigrationFileMetadata(
                filename=file.metadata.filename,
                version=file.metadata.version,
                description=file.metadata.description,
                checksum=self.calculate_checksum(resolved_contents),
            ),
            contents=resolved_contents,
        )

    @staticmethod
    def _resolve_path(directory: str) -> Path:
        if directory.startswith("file:"):
            return Path(directory.removeprefix("file:"))
        raise ValueError(f"unsupported migration directory scheme in '{directory}'. supported schemes: 'file:'")

    @staticmethod
    def _read_raw(path: Path, match: re.Match[str]) -> MigrationFile:
        data: dict[str, Any] = yaml.safe_load(path.read_text())
        return MigrationFile(
            metadata=MigrationFileMetadata(
                filename=path.name,
                version=match.group(1),
                description=match.group(3),
                checksum=0,
            ),
            contents=MigrationFileContents(requests=[MigrationFileLoader._parse_request(r) for r in data["requests"]]),
        )

    @staticmethod
    def _parse_request(raw: dict[str, Any]) -> MigrationFileRequestDefinition:
        return MigrationFileRequestDefinition(
            method=str(raw["method"]),
            path=str(raw["path"]),
            content_type=str(raw["contentType"]) if "contentType" in raw else None,
            params={str(k): str(v) for k, v in raw["params"].items()} if "params" in raw else None,
            body=str(raw["body"]) if "body" in raw else None,
        )

    @staticmethod
    def _remove_nulls(obj: Any) -> Any:
        if isinstance(obj, dict):
            d = cast(dict[str, Any], obj)
            return {k: MigrationFileLoader._remove_nulls(v) for k, v in d.items() if v is not None}
        if isinstance(obj, list):
            lst = cast(list[Any], obj)
            return [MigrationFileLoader._remove_nulls(item) for item in lst]
        return obj

    @staticmethod
    def calculate_checksum(contents: MigrationFileContents) -> int:
        data = {"requests": [r.to_canonical_dict() for r in contents.requests]}
        canonical = json.dumps(
            MigrationFileLoader._remove_nulls(data),
            sort_keys=True,
            separators=(",", ":"),
            ensure_ascii=False,
        ).encode("utf-8")
        digest = hashlib.md5(canonical, usedforsecurity=False).digest()
        return struct.unpack(">i", digest[:4])[0]

File: implementations/typescript/src/migration/loader.ts
import { createHash } from "node:crypto";
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";
import type { MigrationFile, MigrationFileContents, MigrationFileRequestDefinition } from "./model.js";
import { compareMigrationFiles, toCanonicalDict } from "./model.js";
import type { MigrationTemplateResolver } from "./template.js";

// Canonical JSON: keys sorted alphabetically at every level, null/undefined object values
// dropped, compact separators. Matches Python's
// json.dumps(remove_nulls(data), sort_keys=True, separators=(",", ":"), ensure_ascii=False).
// Only ever fed the closed shape produced by toCanonicalDict — do not reuse for arbitrary JSON.
function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJson(item)).join(",")}]`;
  }
  const entries = Object.entries(value as Record<string, unknown>)
    .filter((entry) => entry[1] !== null && entry[1] !== undefined)
    .sort(([a], [b]) => (a < b ? -1 : 1));
  return `{${entries.map(([k, v]) => `${JSON.stringify(k)}:${canonicalJson(v)}`).join(",")}}`;
}

// Canonical algorithm (identical across all implementations):
// canonical JSON of {"requests": [...]} → UTF-8 → MD5 → first 4 bytes as big-endian signed int32.
export function calculateChecksum(contents: MigrationFileContents): number {
  const data = { requests: contents.requests.map((r) => toCanonicalDict(r)) };
  const digest = createHash("md5").update(canonicalJson(data), "utf8").digest();
  return digest.readInt32BE(0);
}

const FILE_NAME_PATTERN = /^V((\d+\.?)+)__(\w+)\.yml$/;

export class MigrationFileLoader {
  private readonly migrationDirectory: string;
  private readonly templateResolver: MigrationTemplateResolver;

  constructor(migrationDirectory: string, templateResolver: MigrationTemplateResolver) {
    this.migrationDirectory = migrationDirectory;
    this.templateResolver = templateResolver;
  }

  load(): MigrationFile[] {
    const dir = resolveDirectoryPath(this.migrationDirectory);
    const rawFiles: MigrationFile[] = [];
    for (const name of readdirSync(dir)) {
      const filePath = join(dir, name);
      if (!statSync(filePath).isFile()) {
        continue;
      }
      const match = FILE_NAME_PATTERN.exec(name);
      if (match === null) {
        continue;
      }
      rawFiles.push(readRawFile(filePath, name, match));
    }
    rawFiles.sort(compareMigrationFiles);
    this.templateResolver.validate(rawFiles);
    return rawFiles.map((file) => this.resolveFile(file));
  }

  private resolveFile(file: MigrationFile): MigrationFile {
    const resolvedContents = this.templateResolver.resolveContents(file.contents);
    return {
      metadata: { ...file.metadata, checksum: calculateChecksum(resolvedContents) },
      contents: resolvedContents,
    };
  }
}

function resolveDirectoryPath(directory: string): string {
  if (directory.startsWith("file:")) {
    return directory.slice("file:".length);
  }
  throw new Error(`unsupported migration directory scheme in '${directory}'. supported schemes: 'file:'`);
}

function readRawFile(filePath: string, filename: string, match: RegExpExecArray): MigrationFile {
  const version = match[1];
  const description = match[3];
  if (version === undefined || description === undefined || version.split(".").some((s) => s === "")) {
    // Empty segment (e.g. trailing dot in "V1.__X.yml") — fail loud, mirroring Python's
    // int("") ValueError, instead of silently treating it as 0.
    throw new Error(`invalid migration filename: ${filename}`);
  }
  const data = parse(readFileSync(filePath, "utf8")) as { requests: Record<string, unknown>[] };
  return {
    metadata: { filename, version, description, checksum: 0 },
    contents: { requests: data.requests.map((raw) => parseRequest(raw)) },
  };
}

function parseRequest(raw: Record<string, unknown>): MigrationFileRequestDefinition {
  return {
    method: String(raw.method),
    path: String(raw.path),
    contentType: "contentType" in raw ? String(raw.contentType) : null,
    params:
      "params" in raw
        ? Object.fromEntries(
            Object.entries(raw.params as Record<string, unknown>).map(([k, v]) => [String(k), String(v)]),
          )
        : null,
    body: "body" in raw ? String(raw.body) : null,
  };
}
